File: js/cloud.js
// cloud.js — optional account + cloud sync. Sign in with an emailed link (no passwords),
// and the whole garden saves to the database on every change. Zero dependencies: raw
// fetch against Supabase's auth + REST endpoints.
import { store } from './store.js';
import { toast } from './ui.js';

// ── Supabase project. Leave blank and Bloom stays fully local — the account
// section simply hides itself. The publishable key is designed to be public.
export const CLOUD = {
  url: 'https://rbljgkthmbfvqtataocc.supabase.co',
  key: 'sb_publishable__JUTOmE75UnmBQav7J9CCw_UbGu4iY8',
};

const SKEY = 'bloom.session.v1';
export const cloudConfigured = () => !!(CLOUD.url && CLOUD.key);

let session = null;
try { session = JSON.parse(localStorage.getItem(SKEY) || 'null'); } catch { session = null; }
const saveSession = (s) => {
  session = s;
  if (s) localStorage.setItem(SKEY, JSON.stringify(s));
  else localStorage.removeItem(SKEY);
};

export const signedIn = () => !!session?.access_token;
export const userEmail = () => session?.user?.email || null;

// ── sync status (Settings listens to show "synced just now" etc.)
let status = 'idle'; // idle | syncing | synced | error
const statusListeners = new Set();
export const syncStatus = () => status;
export const onSyncStatus = (fn) => { statusListeners.add(fn); return () => statusListeners.delete(fn); };
const setStatus = (s) => { status = s; for (const fn of [...statusListeners]) fn(s); };

const jsonHeaders = (authed) => ({
  'Content-Type': 'application/json',
  apikey: CLOUD.key,
  // publishable keys aren't JWTs — only send Authorization once we hold a user token
  ...(authed && session ? { Authorization: `Bearer ${session.access_token}` } : {}),
});

// ── auth: email → sign-in link → the link lands back on Bloom with a session in the hash
export async function requestLink(email) {
  // redirect_to brings the link back to THIS site (github.io / vercel / localhost) —
  // each origin must be allowlisted in Supabase Auth → URL Configuration → Redirect URLs
  const back = encodeURIComponent(location.origin + location.pathname);
  const res = await fetch(`${CLOUD.url}/auth/v1/otp?redirect_to=${back}`, {
    method: 'POST', headers: jsonHeaders(false),
    body: JSON.stringify({ email, create_user: true }),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.msg || err.error_description || 'Could not send the link');
  }
}

// The email carries a 6-digit code too (once the template includes {{ .Token }}) —
// typing it signs in right here, no link-click needed.
export async function verifyCode(email, token) {
  const res = await fetch(`${CLOUD.url}/auth/v1/verify`, {
    method: 'POST', headers: jsonHeaders(false),
    body: JSON.stringify({ email, token, type: 'email' }),
  });
  const data = await res.json().catch(() => ({}));
  if (!res.ok || !data.access_token) throw new Error(data.msg || data.error_description || 'Wrong or expired code');
  saveSession(data);
  await firstSync();
}

const decodeJwt = (t) => {
  try { return JSON.parse(atob(t.split('.')[1].replace(/-/g, '+').replace(/_/g, '/'))); }
  catch { return null; }
};

// Call once at boot, before routing: if the URL hash carries Supabase auth tokens
// (the user just clicked their sign-in link), turn them into a session and clean the URL.
export function handleAuthRedirect() {
  const h = location.hash || '';
  if (!/access_token=|error_description=/.test(h)) return false;
  const params = new URLSearchParams(h.replace(/^#\/?/, ''));
  const clean = () => history.replaceState(null, '', location.pathname + location.search);
  const errDesc = params.get('error_description');
  if (errDesc) {
    clean();
    toast(errDesc.replace(/\+/g, ' '), 'leaf');
    return false;
  }
  const access_token = params.get('access_token');
  const refresh_token = params.get('refresh_token');
  const claims = decodeJwt(access_token);
  clean();
  if (!access_token || !claims?.sub) return false;
  saveSession({ access_token, refresh_token, user: { id: claims.sub, email: claims.email } });
  toast(`Signed in as ${claims.email}`, 'flower');
  return true;
}

async function refreshSession() {
  if (!session?.refresh_token) return false;
  try {
    const res = await fetch(`${CLOUD.url}/auth/v1/token?grant_type=refresh_token`, {
      method: 'POST', headers: jsonHeaders(false),
      body: JSON.stringify({ refresh_token: session.refresh_token }),
    });
    const data = await res.json().catch(() => ({}));
    if (res.ok && data.access_token) { saveSession(data); return true; }
    // sign out only when the token is definitively dead — hiccups keep you signed in
    if ([400, 401, 403].includes(res.status)) saveSession(null);
    return false;
  } catch { return false; } // offline — stay signed in, try again next time
}

async function authedFetch(path, opts = {}, retry = true) {
  const res = await fetch(`${CLOUD.url}${path}`, {
    ...opts, headers: { ...jsonHeaders(true), ...(opts.headers || {}) },
  });
  if (res.status === 401 && retry && await refreshSession()) return authedFetch(path, opts, false);
  return res;
}

export function signOut() {
  authedFetch('/auth/v1/logout', { method: 'POST' }).catch(() => {});
  saveSession(null);
  setStatus('idle');
  toast('Signed out — this browser keeps its local copy', 'leaf');
}

// ── garden sync: one row per user, whole garden as JSON, newest edit wins
export async function pushGarden() {
  if (!signedIn()) return;
  setStatus('syncing');
  try {
    const res = await authedFetch('/rest/v1/gardens', {
      method: 'POST', headers: { Prefer: 'resolution=merge-duplicates' },
      body: JSON.stringify({
        user_id: session.user.id,
        data: store.state,
        edited_at: store.state.editedAt || new Date().toISOString(),
      }),
    });
    setStatus(res.ok ? 'synced' : 'error');
  } catch { setStatus('error'); }
}

async function pullGarden() {
  const res = await authedFetch(`/rest/v1/gardens?select=data,edited_at&user_id=eq.${session.user.id}`);
  if (!res.ok) throw new Error('pull failed');
  const rows = await res.json();
  return rows[0] || null;
}

// how much real life a garden holds — an empty fresh browser must never outrank a real garden
const substance = (s) => (s?.skills?.length || 0) + (s?.sessions?.length || 0)
  + (s?.tasks?.length || 0) + (s?.notes?.length || 0) + (s?.events?.length || 0)
  + (s?.settings?.name ? 1 : 0); // even just a name is worth keeping

// after sign-in (or on boot): a garden with actual plants beats an empty one,
// no matter the timestamps; only between two real gardens does newest-edit win
async function firstSync() {
  setStatus('syncing');
  const row = await pullGarden().catch(() => null);
  const cloudHas = substance(row?.data);
  const localHas = substance(store.state);
  const cloudNewer = row?.data && (row.edited_at || '') > (store.state.editedAt || '');
  if (row?.data && cloudHas && (!localHas || cloudNewer)) {
    store.replace(row.data);
    toast('Your garden is back', 'flower');
    setStatus('synced');
  } else if (!localHas && cloudNewer) {
    setStatus('synced'); // both empty, cloud current — nothing worth writing either way
  } else {
    await pushGarden();
  }
}

let pushT = null;
const queuePush = () => {
  if (!signedIn()) return;
  clearTimeout(pushT);
  pushT = setTimeout(() => pushGarden(), 1500);
};

export function initCloud() {
  if (!cloudConfigured()) return;
  store.setOnSave(queuePush); // every save (silent or not) syncs shortly after
  // refresh the session on every visit so signing in is a once-per-device thing
  if (signedIn()) refreshSession().finally(() => firstSync().catch(() => setStatus('error')));
}

File: ios/Bloom/Cloud.swift
// Cloud.swift — cloud.js ported: email-code sign-in against Supabase auth, whole-garden
// sync to the `gardens` table, newest-edit-wins, substance-guarded first sync.
// Same project, same endpoints, same rules — one garden across web and iPhone.
import Foundation
import Observation

enum Cloud {
    static let url = "https://rbljgkthmbfvqtataocc.supabase.co"
    static let key = "sb_publishable__JUTOmE75UnmBQav7J9CCw_UbGu4iY8"
    static var configured: Bool { !url.isEmpty && !key.isEmpty }
}

struct CloudSession: Codable {
    var access_token: String
    var refresh_token: String?
    var userId: String
    var email: String?
}

enum SyncStatus: String {
    case idle, syncing, synced, error
}

@Observable
final class CloudSync {
    static let shared = CloudSync()

    private(set) var session: CloudSession?
    private(set) var status: SyncStatus = .idle
    @ObservationIgnored weak var store: AppStore?
    @ObservationIgnored private var pushWork: DispatchWorkItem?

    private static let account = "supabase-session"

    var signedIn: Bool { session?.access_token != nil }
    var userEmail: String? { session?.email }

    init() {
        if let data = Keychain.load(account: Self.account) {
            session = try? JSONDecoder().decode(CloudSession.self, from: data)
        }
    }

    private func saveSession(_ s: CloudSession?) {
        session = s
        if let s, let data = try? JSONEncoder().encode(s) {
            Keychain.save(data, account: Self.account)
        } else {
            Keychain.delete(account: Self.account)
        }
    }

    // MARK: - Boot

    /// Call once at launch: hook saves, refresh the session, reconcile gardens.
    func start(store: AppStore) {
        self.store = store
        store.onSave = { [weak self] in self?.queuePush() }
        guard Cloud.configured, signedIn else { return }
        Task {
            _ = await refreshSession()
            await firstSync()
        }
    }

    // MARK: - Requests

    private func headers(authed: Bool) -> [String: String] {
        var h = ["Content-Type": "application/json", "apikey": Cloud.key]
        // publishable keys aren't JWTs — only send Authorization once we hold a user token
        if authed, let token = session?.access_token { h["Authorization"] = "Bearer \(token)" }
        return h
    }

    private func request(_ path: String, method: String = "GET", authed: Bool, body: Data? = nil, extraHeaders: [String: String] = [:]) async throws -> (Data, HTTPURLResponse) {
        var req = URLRequest(url: URL(string: Cloud.url + path)!)
        req.httpMethod = method
        for (k, v) in headers(authed: authed).merging(extraHeaders, uniquingKeysWith: { _, b in b }) {
            req.setValue(v, forHTTPHeaderField: k)
        }
        req.httpBody = body
        let (data, resp) = try await URLSession.shared.data(for: req)
        return (data, resp as! HTTPURLResponse)
    }

    /// 401 → refresh → retry once, like the web's authedFetch.
    private func authedRequest(_ path: String, method: String = "GET", body: Data? = nil, extraHeaders: [String: String] = [:], retry: Bool = true) async throws -> (Data, HTTPURLResponse) {
        let (data, resp) = try await request(path, method: method, authed: true, body: body, extraHeaders: extraHeaders)
        if resp.statusCode == 401, retry, await refreshSession() {
            return try await authedRequest(path, method: method, body: body, extraHeaders: extraHeaders, retry: false)
        }
        return (data, resp)
    }

    // MARK: - Auth

    struct AuthError: LocalizedError {
        var message: String
        var errorDescription: String? { message }
    }

    private static func serverMessage(_ data: Data) -> String? {
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        return (obj["msg"] ?? obj["error_description"] ?? obj["message"]) as? String
    }

    /// Sends the sign-in email (magic link + code, once the template includes {{ .Token }}).
    func requestCode(email: String) async throws {
        let body = try JSONSerialization.data(withJSONObject: ["email": email, "create_user": true])
        let (data, resp) = try await request("/auth/v1/otp", method: "POST", authed: false, body: body)
        guard (200..<300).contains(resp.statusCode) else {
            throw AuthError(message: Self.serverMessage(data) ?? "Could not send the code")
        }
    }

    /// Types the 6-digit code from the email → session → first sync.
    func verifyCode(email: String, code: String) async throws {
        let body = try JSONSerialization.data(withJSONObject: ["email": email, "token": code, "type": "email"])
        let (data, resp) = try await request("/auth/v1/verify", method: "POST", authed: false, body: body)
        guard (200..<300).contains(resp.statusCode),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let access = obj["access_token"] as? String
        else {
            throw AuthError(message: Self.serverMessage(data) ?? "Wrong or expired code")
        }
        let refresh = obj["refresh_token"] as? String
        let user = obj["user"] as? [String: Any]
        saveSession(CloudSession(
            access_token: access,
            refresh_token: refresh,
            userId: user?["id"] as? String ?? "",
            email: user?["email"] as? String ?? email
        ))
        await firstSync()
    }

    @discardableResult
    private func refreshSession() async -> Bool {
        guard let refresh = session?.refresh_token else { return false }
        do {
            let body = try JSONSerialization.data(withJSONObject: ["refresh_token": refresh])
            let (data, resp) = try await request("/auth/v1/token?grant_type=refresh_token", method: "POST", authed: false, body: body)
            if (200..<300).contains(resp.statusCode),
               let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let access = obj["access_token"] as? String {
                let user = obj["user"] as? [String: Any]
                saveSession(CloudSession(
                    access_token: access,
                    refresh_token: obj["refresh_token"] as? String ?? refresh,
                    userId: user?["id"] as? String ?? session?.userId ?? "",
                    email: user?["email"] as? String ?? session?.email
                ))
                return true
            }
            // sign out only when the token is definitively dead — hiccups keep you signed in
            if [400, 401, 403].contains(resp.statusCode) {
                saveSession(nil)
                await MainActor.run { status = .idle }
            }
            return false
        } catch {
            return false   // offline — stay signed in, try again next time
        }
    }

    func signOut() {
        Task { _ = try? await authedRequest("/auth/v1/logout", method: "POST") }
        saveSession(nil)
        status = .idle
        store?.toast("Signed out — this phone keeps its local copy", "leaf")
    }

    // MARK: - Garden sync (one row per user, whole garden as JSON, newest edit wins)

    private struct PushBody: Encodable {
        var user_id: String
        var data: BloomState
        var edited_at: String
    }

    private struct PullRow: Decodable {
        var data: BloomState
        var edited_at: String?
    }

    func queuePush() {
        guard signedIn else { return }
        pushWork?.cancel()
        let work = DispatchWorkItem { [weak self] in
            Task { await self?.pushGarden() }
        }
        pushWork = work
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.5, execute: work)
    }

    func pushGarden() async {
        guard signedIn, let store else { return }
        await MainActor.run { status = .syncing }
        do {
            let payload = await MainActor.run {
                PushBody(user_id: session?.userId ?? "", data: store.state,
                         edited_at: store.state.editedAt ?? nowISO())
            }
            let body = try JSONEncoder().encode(payload)
            let (_, resp) = try await authedRequest("/rest/v1/gardens", method: "POST", body: body,
                                                    extraHeaders: ["Prefer": "resolution=merge-duplicates"])
            await MainActor.run { status = (200..<300).contains(resp.statusCode) ? .synced : .error }
        } catch {
            await MainActor.run { status = .error }
        }
    }

    private func pullGarden() async -> PullRow? {
        guard let uid = session?.userId else { return nil }
        guard let (data, resp) = try? await authedRequest("/rest/v1/gardens?select=data,edited_at&user_id=eq.\(uid)"),
              (200..<300).contains(resp.statusCode),
              let rows = try? JSONDecoder().decode([PullRow].self, from: data)
        else { return nil }
        return rows.first
    }

    /// After sign-in (or at boot): a garden with actual plants beats an empty one,
    /// no matter the timestamps; only between two real gardens does newest-edit win.
    func firstSync() async {
        guard let store else { return }
        await MainActor.run { status = .syncing }
        let row = await pullGarden()
        await MainActor.run {
            let cloudHas = row?.data.substance ?? 0
            let localHas = store.state.substance
            let cloudNewer = row != nil && (row?.edited_at ?? "") > (store.state.editedAt ?? "")
            if let row, cloudHas > 0, localHas == 0 || cloudNewer {
                store.replace(row.data)
                store.toast("Your garden is back", "flower")
                status = .synced
            } else if localHas == 0 && cloudNewer {
                status = .synced   // both empty, cloud current — nothing worth writing either way
            } else {
                Task { await self.pushGarden() }
            }
        }
    }
}
